Allow to disable SQLite schema emulation in SqliteSchemaManager

mvorisek opened the pull request:
|      Q       |   A
|------------- | -----------
| Type         | bug
| Fixed issues | #6314

#### Summary

Missed in https://github.com/doctrine/dbal/pull/4804 and fix related https://github.com/doctrine/dbal/pull/5517.

When merging up (into 4.x), simply remove the `str_replace`.

derrabus commented:
Can we cover this with a test please?

mvorisek commented:
Is your request for a one functional test into `Doctrine\DBAL\Tests\Functional\Schema\SqliteSchemaManagerTest`?

mvorisek commented:
@derrabus may I understand your wishes better?

derrabus commented:
A test please.

mvorisek commented:
@greg0ire can this bugfix be merged? It contains a functional test as requested.

greg0ire commented on tests/Functional/Schema/SqliteSchemaManagerTest.php:
```diff
@@ -371,4 +374,57 @@ public function testShorthandInForeignKeyReferenceWithMultipleColumns(): void
+    {
+        $this->dropTableIfExists('`list_table_no_schema_emulation.test`');
+
+        $ddl = <<<'DDL'
```
Let's not clutter the reader's mind with one-time, inconsistently named variables.

```suggestion
        $this->connection->executeStatement(<<<'DDL'
```

mvorisek replied:
done

greg0ire commented on tests/Functional/Schema/SqliteSchemaManagerTest.php:
```diff
@@ -371,4 +374,57 @@ public function testShorthandInForeignKeyReferenceWithMultipleColumns(): void
+            FOREIGN KEY (parent_id) REFERENCES `list_table_no_schema_emulation.test` (id)
+        );
+        DDL;
+        $dd2 = <<<'DDL'
```
```suggestion
        $this->connection->executeStatement(<<<'DDL'
```

mvorisek replied:
done

greg0ire commented on tests/Functional/Schema/SqliteSchemaManagerTest.php:
```diff
@@ -371,4 +374,57 @@ public function testShorthandInForeignKeyReferenceWithMultipleColumns(): void
+        CREATE INDEX i ON `list_table_no_schema_emulation.test` (parent_id);
+        DDL;
+
+        $this->connection->executeStatement($ddl);
+        $this->connection->executeStatement($dd2);
```
```suggestion
```

mvorisek replied:
done

greg0ire commented on tests/Functional/Schema/SqliteSchemaManagerTest.php:
```diff
@@ -371,4 +374,57 @@ public function testShorthandInForeignKeyReferenceWithMultipleColumns(): void
+
+        $schemaManager = $this->schemaManager;
+        $refl          = new ReflectionMethod($schemaManager, 'selectTableColumns');
+        $refl->setAccessible(true);
```
Changing method accessibility in tests is a smell. Why do you need to do it?

mvorisek replied:
This is because full table introspection with schema is not supported (yet).

Therefore, I test functional behaviour using protected methods.

mvorisek replied:
addressed

mvorisek commented:
The CI is passing.

mvorisek commented:
@derrabus may I ask you for a review?

derrabus commented:
Sure. It's on my todo list, but I don't have much time to spend on open source work at the moment.

derrabus commented:
I don't quite understand this test, tbh. You're testing with a dot in a table name that SQLite to my astonishment seems to accept when quoted. But that doesn't have much to do with schema emulation. Shouldn't the test work with a table from an actual secondary schema?

mvorisek commented:
The fixed methods accept database name using 1st parameter, the table name is accepted as 2nd parameter, thus the 2nd parameter can be unquoted and dot must be preserved (it cannot contain schema).

As the fixed methods cannot be reached easily from public API as table with schema and dot is not fully supported yet, the test is coded to call the fixed protected methods.

derrabus commented:
> As the fixed methods cannot be reached easily from public API as table with schema and dot is not fully supported yet, the test is coded to call the fixed protected methods.

So, what does this PR fix then? An implementation detail?

mvorisek commented:
The schema manager can be extended by user classes and the protected methods must behave as expected. This PR fixes the behaviour and it is tested with functional test. For more details, see the #4804 and #5517 PRs, ie. the schema emulation is deprecated for 3.x and removed in 4.x. Can we agree on this? Is there anything else I should do?

mvorisek commented:
Can this PR be merged, please.

mvorisek commented:
Can this PR be merged, please?

mvorisek commented:
@derrabus can I have your review, please?

mvorisek commented:
Hi @derrabus, we are waiting for your approval. The "SQLite schema emulation to manager" was added in #5268 without any explanation. I belive it is there for historical reasons and such emulation was officially removed in #4804 (for v4.x) and in #5517 (for v3.x with the possibility to disable it).

Given these facts, I belive the emulation simply should not be there, ie. it should be disableable for v3.x and removed for v4.x.

Tests are testing the desired behaviour.

My question to you is, what can I do more to get this PR merged and if we are on the same page with the reasoning why it should be removed. Thank you!

mvorisek commented:
Can this PR be please either approved or can I get an actionable feedback?

mvorisek commented:
@greg0ire there are 2 approving reviews, can this PR be merged?

mvorisek commented:
Can this PR be merged?

derrabus commented:
I'm working on my DBAL backlog as I find the time to do so. I was able to clear some items from that list today and will come back to this PR soon. Bear with me.

mvorisek commented:
This PR is about 2 month since approved... Can it be merged now, please.

derrabus commented:
I'm sorry, this took me so long. But my concern remains: your test runs assertions on implementation details. It looks complicated and I don't want to maintain it. Your change looks like you're fixing a valid problem, but I need a functional test (no reflection voodoo) that reproduces your problem as you encounter it.

mvorisek commented:
Thank you - please tell me how to test better.

derrabus commented:
Write a test that reproduces your problem as you encounter it. That's all I can tell you because I have no clue why you're hitting this issue at all.

mvorisek commented:
We override non-final SqliteSchemaManager class. I propose to merge this as is as in v4.0 the schema emulation has been removed and thus I would be fine with empty merge/test when merging up.

derrabus commented:
> We override non-final SqliteSchemaManager class.

So, do the same in the test.

mvorisek commented:
@derrabus addressed

mvorisek commented:
> > We override non-final SqliteSchemaManager class.
> 
> So, do the same in the test.

@derrabus done

mvorisek commented:
@derrabus the test was adjusted, can I have your final review?

derrabus commented:
I'm touring conferences this week. Please ping me again if I haven't reviewed the PR by the end of next week.

mvorisek commented:
@derrabus ping as requested

mvorisek commented:
@derrabus pinging you again as requested

mvorisek commented:
TODO for myself - rebase

mvorisek commented:
@derrabus rebased and pinging you as requested, I hope you enjoyed the conferences!

mvorisek commented:
> I'm touring conferences this week. Please ping me again if I haven't reviewed the PR by the end of next week.

@derrabus pinging you again :)

mvorisek commented:
@derrabus pinging you again

mvorisek commented:
@greg0ire might I have your review? @derrabus requested me to ping him, but he seems not responsive, not sure what happened. I would be happy if we can move this PR forward. Thank you!

greg0ire commented:
Please squash your commits. Once this is merged, nobody needs to know you wrote non functional tests at some point.

While squashing, please improve your commit message according to [the contributing guide](https://www.doctrine-project.org/contribute/index.html#crafting-meaningful-commit-messages).

Right now it says that you "fix something unwanted"… unwanted stuff is avoided, not fixed. More importantly, you should state why you're doing this. Reading the test, I can see that you need to have a custom schema manager to even hit the bug, so this warrants an explanation, it's not obvious this only ever causes issue when using a custom schema manager

greg0ire commented on tests/Functional/Schema/SqliteSchemaManagerTest.php:
```diff
@@ -398,6 +400,82 @@ public function testShorthandInForeignKeyReferenceWithMultipleColumns(): void
+                static fn (array $row) => [$row['table_name'], $row['name']],
+                $customSqliteSchemaManager->selectTableColumnsWithSchema(),
+            ),
+        );
```
To reviewers: only this assertion fails without the bugfix… it fails like so:

```
1) Doctrine\DBAL\Tests\Functional\Schema\SqliteSchemaManagerTest::testListTableNoSchemaEmulation
Failed asserting that two arrays are identical.
--- Expected
+++ Actual
@@ @@
-Array &0 (
-    0 => Array &1 (
-        0 => 'list_table_no_schema_emulation.test'
-        1 => 'id'
-    )
-    1 => Array &2 (
-        0 => 'list_table_no_schema_emulation.test'
-        1 => 'parent_id'
-    )
-)
+Array &0 ()
```

mvorisek committed:
Fix SQLite schema emulation in SqliteSchemaManager

The schema emulation is a deprecated feature.

The SqlitePlatform does support disabling the schema emulation already, the SqliteSchemaManager should support it too to ease the migration path between 3.x and 4.x.

The test requires custom SqliteSchemaManager because the current SqliteSchemaManager  impl. does not support tables with schema yet.

mvorisek commented:
Squashed and repushed.

greg0ire reviewed (approved):
Great commit message, please write them like this from now on.

A concern I have is that this change is here to support a feature that can only be enabled by overriding the schema manager, but why not… there is not much harm IMO since it will be gone in 4.0 (right?).

greg0ire commented:
@derrabus @SenseException please review

greg0ire commented:
Thanks @mvorisek !
